lib/pengine/rules.c

```diff
@@ -965,12 +965,163 @@ pe_eval_subexpr(xmlNode *expr, pe_rule_eval_data_t *rule_data, crm_time_t *next_
+                          l_val, r_val);
+                cmp = compare_attr_expr_vals(l_val, r_val, "string", op);
+            }
+
+        } else if (pcmk__str_eq(type, "number", pcmk__str_casei)) {
+            double l_val_num;
+            double r_val_num;
+
+            int rc1 = pcmk__scan_double(l_val, &l_val_num, NULL, NULL);
+            int rc2 = pcmk__scan_double(r_val, &r_val_num, NULL, NULL);
+
+            if (rc1 == pcmk_rc_ok && rc2 == pcmk_rc_ok) {
+                if (l_val_num < r_val_num) {
+                    cmp = -1;
+                } else if (l_val_num > r_val_num) {
+                    cmp = 1;
+                } else {
+                    cmp = 0;
+                }
+
+            } else {
+                crm_debug("Floating-point parse error. Comparing %s and %s as "
+                          "strings", l_val, r_val);
+                cmp = compare_attr_expr_vals(l_val, r_val, "string", op);
+            }
+
+        } else if (pcmk__str_eq(type, "version", pcmk__str_casei)) {
+            cmp = compare_version(l_val, r_val);
+
+        }
+
+    } else if (l_val == NULL && r_val == NULL) {
+        cmp = 0;
+    } else if (r_val == NULL) {
+        cmp = 1;
+    } else {    // l_val == NULL && r_val != NULL
+        cmp = -1;
+    }
+
+    return cmp;
+}
+
+/*!
+ * \internal
+ * \brief   Check whether an attribute expression evaluates to \c true
+ *
+ * \param[in]   l_val   Value on left-hand side of comparison
+ * \param[in]   r_val   Value on right-hand side of comparison
+ * \param[in]   type    How to interpret the values (allowed values:
+ *                      "string", "integer", "number", "version",
+ *                      \c NULL)
+ * \param[in]   op      Type of comparison.
+ *
+ * \return  \c true if expression evaluates to \c true, \c false
+ *          otherwise
+ */
+static bool
+accept_attr_expr(const char *l_val, const char *r_val, const char *type,
+                 const char *op)
+{
+    int cmp;
+
+    if (pcmk__str_eq(op, "defined", pcmk__str_casei)) {
+        return (l_val != NULL);
+
+    } else if (pcmk__str_eq(op, "not_defined", pcmk__str_casei)) {
+        return (l_val == NULL);
+    }
+
+    cmp = compare_attr_expr_vals(l_val, r_val, type, op);
+
+    if (pcmk__str_eq(op, "eq", pcmk__str_casei)) {
+        return (cmp == 0);
+
+    } else if (pcmk__str_eq(op, "ne", pcmk__str_casei)) {
+        return (cmp != 0);
+
+    } else if (l_val == NULL || r_val == NULL) {
+        // The comparison is meaningless from this point on
+        return FALSE;
+
+    } else if (pcmk__str_eq(op, "lt", pcmk__str_casei)) {
+        return (cmp < 0);
+
+    } else if (pcmk__str_eq(op, "lte", pcmk__str_casei)) {
+        return (cmp <= 0);
+
+    } else if (pcmk__str_eq(op, "gt", pcmk__str_casei)) {
+        return (cmp > 0);
+
+    } else if (pcmk__str_eq(op, "gte", pcmk__str_casei)) {
+        return (cmp >= 0);
+    }
+
+    return false;   // Should never reach this point
+}
+
```

Comment:
May want to break this into two commits (one to refactor into separate functions, and one to do the type logic fix).

Reply:
The push just now is the same content as the one this afternoon, but with the libpe_status commit broken into three commits.